Web Quiz Engine (Kotlin)/task/src/engine/service/QuizService.kt

```diff
@@ -53,14 +54,34 @@ class QuizService(
             }
         }
 
+    fun findCompletedQuizPageForCurrentUser(pageable: Pageable): Page<QuizCompletionDto> =
+        quizCompletionRepository.findAllByUserUserIdAndQuizIdIsNotNullAndCompletedAtIsNotNull(
+            userId = SecurityUtil.authenticatedUser().userId!!,
+            pageable = pageable,
+            ).map { quizCompletion ->
+            with(quizCompletion) {
+                QuizCompletionDto(
+                    id = quiz.id!!, // nullable id is filtered out by the database query
+                    completedAt = completedAt!! //same
+                )
+            }
+        }
+
     fun checkQuizSolution(quizId: Int, answerList: List<Int>?): QuizPostResponseBody =
         findQuizByIdOrThrow(quizId)
             .let { quiz ->
             if ((quiz.answerList.isNullOrEmpty() && answerList.isNullOrEmpty()) ||
                 quiz.answerList?.size == answerList?.size &&
                 (quiz.answerList?.containsAll(answerList.orEmpty()) == true &&
-                        answerList?.containsAll(quiz.answerList) == true))
+                        answerList?.containsAll(quiz.answerList) == true)) {
```

Comment:
I thought we had removed this duplicate execution of `containsAll`.
You don't need to do a reverse `containsAll` after you've already done the size comparison and a `containsAll` between the two lists.

Reply:
I had done only half the job 😅 Thks for your eagle eye.

Reply:
Done as
```kotlin
if ((quiz.answerList.isNullOrEmpty() && answerList.isNullOrEmpty()) ||
quiz.answerList?.size == answerList?.size &&
quiz.answerList?.containsAll(answerList.orEmpty()) == true) 
```